photutils/psf/groupstars.py

```diff
@@ -0,0 +1,228 @@
+                             'real number. Got {}'.format(crit_separation))
+        else:
+            self._crit_separation = crit_separation
+ 
+    def group_stars(self, starlist):
+        cstarlist = starlist.copy()
+
+        if 'id' not in cstarlist.colnames:
+            cstarlist.add_column(Column(name='id',
+                                        data=np.arange(len(cstarlist)) + 1))
+        cstarlist.add_column(Column(name='group_id',\
+                                data=np.zeros(len(cstarlist), dtype=np.int)))
+
+        if not np.array_equal(cstarlist['id'], np.arange(len(cstarlist)) +1):
+            raise ValueError('id colum must be an integer-valued ' +
+                             'sequence starting from 1. ' +
+                             'Got {}'.format(cstarlist['id']))
+
+        n = 1
+        while (cstarlist['group_id'] == 0).sum() > 0:
+            init_star = cstarlist[np.where(cstarlist['group_id'] == 0)[0][0]]
+            index = self.find_group(init_star,
+                                    cstarlist[cstarlist['group_id'] == 0])
+            cstarlist['group_id'][index-1] = n
+            k = 1
+            K = len(index)
+            while k < K:
+                init_star = cstarlist[cstarlist['id'] == index[k]]
+                tmp_index = self.find_group(init_star,\
+                                        cstarlist[cstarlist['group_id'] == 0])
+                if len(tmp_index) > 0:
+                    cstarlist['group_id'][tmp_index-1] = n
+                    index = np.append(index, tmp_index)
+                    K = len(index)
+                k += 1
+            n += 1
+
+        return cstarlist
+
+    def find_group(self, star, starlist):
+        """
+        Find the ids of those stars in ``starlist`` which are at a distance less
+        than ``crit_separation`` from ``star``.
+
+        Parameters
+        ----------
+        star : `~astropy.table.Row`
+            Star which will be either the head of a cluster or an isolated one.
+        starlist : `~astropy.table.Table`
+            List of stars positions. Columns named as ``x_0`` and ``y_0``,
+            which corresponds to the centroid coordinates of the sources,
+            must be provided.
+        
+        Returns
+        -------
+        Array containing the ids of those stars which are at a distance less
+        than ``crit_separation`` from ``star``.
+        """
+        
+        star_distance = np.hypot(star['x_0'] - starlist['x_0'],
+                                 star['y_0'] - starlist['y_0'])
+        distance_criteria = star_distance < self.crit_separation
+        return np.asarray(starlist[distance_criteria]['id'])
+
+
+class DBSCANGroup(GroupStarsBase):
+    """
+    This class creates groups stars according to a distance criteria using the
+    Density-based Spatial Clustering of Applications with Noise (DBSCAN) from
+    scikit-learn.
```

Comment:
This is not critical if you think it will be too complex to explain (or just don't know the answer), but it might be useful to add a note to saying when this particular algorithm might be useful.